### PR TITLE
[IMP] pos_restuarant: orderlines should stick to tables during table unmerge

### PR DESCRIPTION
**Pull request by jipr-odoo**
Before this commit:
===================
- If items were ordered on Table 1 and then merged into Table 2, those items
  had to be reordered for Table 2.
- When unmerging Table 1 from Table 2, Table 1 would lose its original orders,
  resulting in an empty table.
- A traceback occurred when attempting to merge more than two tables, each
  containing orders.

After this commit:
=============
- Items ordered on Table 1 remain ordered when merged into Table 2, eliminating
  the need to reorder.
- When unmerging, Table 1 regains its original orders. Items retain their status
  as ordered or not ordered based on their state(ordered or unorderd).
- The traceback issue when merging more than two tables with existing orders
  has been resolved.

Task- 4224207

Related Enterprise PR: https://github.com/odoo/enterprise/pull/74342

### CODE REVIEW

File: addons/pos_restaurant/models/pos_restaurant.py [hunk 1]
```diff
@@ -111,6 +111,7 @@ class RestaurantTable(models.Model):
     seats = fields.Integer('Seats', default=1, help="The default number of customer served at this table.")
     color = fields.Char('Color', help="The table's color, expressed as a valid 'background' CSS property value", default="#35D374")
     parent_id = fields.Many2one('restaurant.table', string='Parent Table', help="The parent table if this table is part of a group of tables")
+    child_ids = fields.One2many('restaurant.table', 'parent_id', string='Children Tables')
```
Comment (caburj) [inline addons/pos_restaurant/models/pos_restaurant.py:114]: No need to introduce this field. You can do `table["<-restaurant.table.parent_id"]` in JS to get the children tables.

File: addons/point_of_sale/static/src/app/models/pos_order.js [hunk 2]
```diff
@@ -269,7 +269,14 @@ export class PosOrder extends Base {
...
     hasSkippedChanges() {
-        return this.lines.find((orderline) => orderline.skip_change) ? true : false;
+        return this.lines.find(
+            (orderline) =>
+                orderline.skip_change &&
+                !orderline.uiState.hideSkipChangeClass &&
+                !orderline.origin_order_id
+        )
+            ? true
+            : false;
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/models/pos_order.js:279]:
```suggestion
        return Boolean(this.lines.find(
            (orderline) =>
                orderline.skip_change &&
                !orderline.uiState.hideSkipChangeClass &&
                !orderline.origin_order_id
        ));
```

File: addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js [hunk 3]
```diff
@@ -141,9 +141,11 @@ export class FloorScreen extends Component {
                 table.position_h = table.getX();
                 table.position_v = table.getY();
                 if (table.parent_id) {
-                    this.pos.data.write("restaurant.table", [table.id], {
-                        parent_id: null,
+                    this.unMergeTable(table);
+                    this.pos.data.write("restaurant.table", [table.parent_id.id], {
+                        child_ids: [[3, table.id]],
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:146]: even if you keep `child_ids`, you don't need to do this write.

File: addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js [hunk 4]
```diff
@@ -217,12 +219,15 @@ export class FloorScreen extends Component {
...
-                }
                 this.pos.data.write("restaurant.table", [table.id], {
                     parent_id: this.state.potentialLink.parent.id,
                 });
+                this.pos.data.write("restaurant.table", [table.parent_id.id], {
+                    child_ids: [[4, this.state.potentialLink.child.id]],
+                });
+                if (oToTrans) {
+                    this.pos.transferOrder(oToTrans.uuid, this.state.potentialLink.parent, true);
+                }
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:230]: why did you move this here?

File: addons/pos_restaurant/static/src/app/screens/payment_screen/payment_screen.js [hunk 5]
```diff
@@ -15,12 +15,36 @@ patch(PaymentScreen.prototype, {
...
+        let tables = this.currentOrder?.table_id?.child_ids || [];
+
+        while (tables.length) {
+            const orders = tables
+                .map((table) =>
+                    this.pos.models["pos.order"].find(
+                        (o) => o.table_id?.id === table.id && !o.finalized
+                    )
+                )
+                .filter(Boolean);
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/payment_screen/payment_screen.js:30]:
```suggestion
            const orders = tables.map((t) => t["<-pos.order.table_id"].filter((o) => !o.finalized)).flat();
```

File: addons/pos_restaurant/static/src/app/screens/payment_screen/payment_screen.js [hunk 6]
```diff
@@ -15,12 +15,36 @@ patch(PaymentScreen.prototype, {
...
+
+            orderToDelete.push(...orders);
+            for (const table of tables) {
+                changedTables.push(table);
+            }
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/payment_screen/payment_screen.js:35]: `changedTables` seems to be unused.

File: addons/pos_restaurant/static/src/app/screens/payment_screen/payment_screen.js [hunk 7]
```diff
@@ -15,12 +15,36 @@ patch(PaymentScreen.prototype, {
...
+
         // After the order has been validated the tables have no reason to be merged anymore.
-        const changedTables = this.pos.models["restaurant.table"]?.filter(
-            (t) => t.parent_id && t.parent_id.id === this.currentOrder.table_id?.id
-        );
         if (changedTables?.length) {
             for (const table of changedTables) {
+                this.pos.data.write("restaurant.table", [table.parent_id.id], {
+                    child_ids: [[3, table.id]],
+                });
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/payment_screen/payment_screen.js:47]: You can get rid of this when you removed `child_ids` field.

File: addons/pos_restaurant/static/src/app/services/pos_store.js [hunk 8]
```diff
@@ -325,33 +331,127 @@ patch(PosStore.prototype, {
...
         await this.setTable(destinationTable);
     },
+    async restoreOrdersToOriginalTable(order, destinationTableOrder) {
+        const orderlines = (destinationTableOrder.lines || []).reduce((acc, orderline) => {
+            if (orderline?.origin_order_id) {
+                const line = this.models["pos.order.line"].getBy("uuid", orderline.uuid);
+                acc.push(line);
+            }
+            return acc;
+        }, []);
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/services/pos_store.js:421]:
```suggestion
        const orderlines = destinationTableOrder.lines.filter((line) => Boolean(line.origin_order_id))
```

File: addons/pos_restaurant/static/src/app/services/pos_store.js [hunk 9]
```diff
@@ -325,33 +331,127 @@ patch(PosStore.prototype, {
...
+
+        for (const orderline of orderlines) {
+            if (orderline?.origin_order_id.id === order.id || order.table_id.child_ids.length) {
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/services/pos_store.js:424]: `|| order.table_id.child_ids.length` I don't understand this condition. Can you explain?

File: addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js [hunk 10]
```diff
@@ -530,6 +535,25 @@ export class FloorScreen extends Component {
         const table = await this.pos.data.create("restaurant.table", [newTableData]);
         return table[0];
     }
+    async unMergeTable(table) {
+        const destinationTable = this.pos.getMainTable(table);
+        const destinationTableOrder = this.pos.getActiveOrdersOnTable(destinationTable)?.[0];
+        const orderToRestore = this.pos.models["pos.order"].filter(
+            (o) =>
+                (o.table_id?.id === table.id || o.uiState.initialTable === table.id) && !o.finalized
+        )[0];
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:544]:
```suggestion
        const orderToRestore = table["<-pos.order.table_id"].filter((o) => !o.finalized)[0];
```

File: addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js [hunk 11]
```diff
@@ -530,6 +535,25 @@ export class FloorScreen extends Component {
         const table = await this.pos.data.create("restaurant.table", [newTableData]);
         return table[0];
     }
+    async unMergeTable(table) {
+        const destinationTable = this.pos.getMainTable(table);
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:539]:
```suggestion
        const mainTable = this.pos.getMainTable(table);
```
rename

File: addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js [hunk 12]
```diff
@@ -530,6 +535,25 @@ export class FloorScreen extends Component {
...
+        const destinationTableOrder = this.pos.getActiveOrdersOnTable(destinationTable)?.[0];
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:540]:
```suggestion
        const mainOrder = this.pos.getActiveOrdersOnTable(destinationTable)?.[0];
```

rename

File: addons/pos_restaurant/static/src/app/services/pos_store.js [hunk 13]
```diff
@@ -325,33 +331,127 @@ patch(PosStore.prototype, {
...
             }
-            await this.deleteOrders([order]);
+            if (!isMergedOrder) {
+                order.isTransferedOrder = true;
+                await this.deleteOrders([order]);
+            }
         }
         await this.setTable(destinationTable);
     },
+    async restoreOrdersToOriginalTable(order, destinationTableOrder) {
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/services/pos_store.js:414]:
```suggestion
    async restoreOrdersToOriginalTable(orderToExtract, mergedOrder) {
```
rename

File: addons/pos_restaurant/static/src/app/services/pos_store.js [hunk 14]
```diff
@@ -325,33 +331,127 @@ patch(PosStore.prototype, {
...
+
+        for (const orderline of orderlines) {
+            if (orderline?.origin_order_id.id === order.id || order.table_id.child_ids.length) {
+                orderline.update({ order_id: order });
+                if (orderline?.origin_order_id.id === order.id) {
+                    if (orderline.skip_change) {
+                        orderline.toggleSkipChange();
+                    }
+                    orderline.origin_order_id = null;
+                }
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/services/pos_store.js:431]:
```suggestion
                orderline.update({ order_id: order, origin_order_id: null });
                if (orderline.skip_change) {
                    orderline.toggleSkipChange();
                }
```

File: addons/pos_restaurant/models/pos_restaurant.py (hunk 1 above)
Comment (jipr-odoo) [reply addons/pos_restaurant/models/pos_restaurant.py:114]: yes optionally we can use `get children()` method in place of new field

File: addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js (hunk 4 above)
Comment (jipr-odoo) [reply addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:230]: actually, we were trying to update the prep card after merging of table so at that time we needed `parent_id` data before `transferOrder` 
but now we don't update prep card. I will keep it the same as it was earlier.

File: addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js (hunk 10 above)
Comment (jipr-odoo) [reply addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:544]: What if destination table was empty while merging so now we will get empty value for `table["<-pos.order.table_id"]`
to handle this we need to use `table.parent_id` in while loop because there may be scenerio
like: table 1(with orderline) merged to table 2(empty) merged to table 3(empty)

in that case 
```
const orderToRestore = this.pos.models["pos.order"].find(
            (o) =>
                (o.table_id?.id === table.id || o.uiState.initialTable === table.id) && !o.finalized
        )[0];
 ```
        
  we can use this

File: addons/pos_restaurant/static/src/app/services/pos_store.js (hunk 9 above)
Comment (jipr-odoo) [reply addons/pos_restaurant/static/src/app/services/pos_store.js:424]: this condition is used to handled this case where 
more than 2 tables merged for example 3 tables with each 1 orderline 
table 1 --> table 2--> table 3.

now unmerge from `table 2` so now 2 orderlines should get back from `table 3`
but with this condition `orderline?.origin_order_id.id === order.id`
only 1 orderline will get back so `order.table_id.child_ids.length` this condition is used to determine 
that `middle table 2` have also child table with his so take that orderlines also.

see ref video: https://www.awesomescreenshot.com/video/34242062?key=fabf47ac594f977a10ae7dbb7a2b2933

Comment (jipr-odoo) [reply addons/pos_restaurant/static/src/app/services/pos_store.js:424]: Also we will need this condition
```js
if (orderline?.origin_order_id.id === orderToExtract.id) {
    if (orderline.skip_change) {
        orderline.toggleSkipChange();
    }
    orderline.origin_order_id = null;
 }
 ```
 
 since we need to update that orderline only whose belongs to the current table (dragged table)

File: addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js (hunk 10 above)
Comment (caburj) [reply addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:544]: If the table being unmerged was originally empty, I think you should do nothing, right?

Comment (caburj) [reply addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:544]: If the main table is empty, are you moving the order (from secondary table) to the main table?

Comment (jipr-odoo) [reply addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:544]: Nope the table being unmerged was has an order but the destination table was empty 
in that case, while merging the the table_id of that order will updated to destination table 

so while unmerge we need to again update that order's table_id to original one

see: https://www.awesomescreenshot.com/video/34247039?key=6fb10e6df5e46e6f1f80d4756abf8b81

File: addons/pos_restaurant/static/src/app/services/pos_store.js [hunk 15]
```diff
@@ -314,7 +314,13 @@ patch(PosStore.prototype, {
             [...el.classList].find((c) => c.includes("tableId")).split("-")[1]
         );
     },
-    async transferOrder(orderUuid, destinationTable) {
+    getMainTable(table) {
+        while (table.parent_id) {
+            table = table.parent_id;
+        }
+        return table;
+    },
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/services/pos_store.js:322]: move this to `restaurant_table.js`.

it should be:

```js
get rootTable() {
    let table = this;
    while (table) {
        table = table.parent_id;
    }
    return table;
}
```

File: addons/pos_restaurant/static/src/app/models/restaurant_table.js [hunk 16]
```diff
@@ -85,5 +85,8 @@ export class RestaurantTable extends Base {
     getName() {
         return this.table_number.toString();
     }
+    get children() {
+        return this.models["restaurant.table"].filter((t) => t.parent_id?.id === this.id);
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/models/restaurant_table.js:89]:
```suggestion
        return this["<-restaurant.table.parent_id"];
```

I tested and this suggestion seems to be working.

File: addons/pos_restaurant/static/src/app/screens/payment_screen/payment_screen.js [hunk 17]
```diff
@@ -15,10 +15,26 @@ patch(PaymentScreen.prototype, {
...
+        let tables = this.currentOrder?.table_id?.children;
+        while (tables.length) {
+            const orders = tables
+                .map((t) => t["<-pos.order.table_id"].filter((o) => !o.finalized))
+                .flat();
+
+            orderToDelete.push(...orders);
+            for (const table of tables) {
+                changedTables.push(table);
+            }
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/payment_screen/payment_screen.js:30]:
```suggestion
            changedTables.push(...tables);
```

File: addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js [hunk 18]
```diff
@@ -530,6 +529,25 @@ export class FloorScreen extends Component {
         const table = await this.pos.data.create("restaurant.table", [newTableData]);
         return table[0];
     }
+    async unMergeTable(table) {
+        const mainTable = this.pos.getMainTable(table);
+        const mainOrder = this.pos.getActiveOrdersOnTable(mainTable)?.[0];
+        const orderToRestore = this.pos.models["pos.order"].find(
+            (o) =>
+                (o.table_id?.id === table.id || o.origin_table_id?.id === table.id) && !o.finalized
+        );
```
Comment (caburj) [inline addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:538]:
```suggestion
        const orderToRestore = table["<-pos.order.origin_table_id"].find((o) => !o.finalized);
```

File: addons/pos_restaurant/static/src/app/services/pos_store.js (hunk 15 above)
Comment (caburj) [reply addons/pos_restaurant/static/src/app/services/pos_store.js:322]: I think you can also extract `getActiveOrdersOnTable` and move it to `restaurant_table.js` model.

```js
get activeOrders() {
    return this["<-pos.order.table_id"].filter((o) => !o.finalized && o.lines.length > 0);
}
```

File: addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js (hunk 18 above)
Comment (jipr-odoo) [reply addons/pos_restaurant/static/src/app/screens/floor_screen/floor_screen.js:538]: this will work only when destination table is empty 
because only at that time we update the table_id of order 

but when both tables have orders then it will not work 

```
const orderToRestore =
            table["<-pos.order.origin_table_id"].find((o) => !o.finalized) ||
            table["<-pos.order.table_id"].find((o) => !o.finalized);
```
in that case we need to use this

File: addons/pos_restaurant/static/src/app/services/pos_store.js (hunk 15 above)
Comment (jipr-odoo) [reply addons/pos_restaurant/static/src/app/services/pos_store.js:322]: with this way When changing the order_id of an orderline in the pos.order.line model, the changes were not reflected correctly. To resolve this, need to add the this line: 
`mergedOrder = this.models["pos.order"].get(mergedOrder.id);`
 to ensure the updated mergedOrder is retrieved
 
 ref: https://github.com/odoo/odoo/pull/188183/files#diff-d2746a4b66902224e35c430b3dcbd8b42124bb03213d750172cb020860f4324dR415

Comment (caburj) [reply addons/pos_restaurant/static/src/app/services/pos_store.js:322]: > with this way When changing the order_id of an orderline in the pos.order.line model, the changes were not reflected correctly. To resolve this, need to add the this line:
mergedOrder = this.models["pos.order"].get(mergedOrder.id);
to ensure the updated mergedOrder is retrieved

I see. I'm working on a branch that actually resolves such issue. This is a bug in the related_models module.

File: addons/pos_restaurant/static/src/app/services/pos_store.js [hunk 19]
```diff
@@ -325,33 +331,117 @@ patch(PosStore.prototype, {
...
             });
             this.setOrder(destinationOrder);
             if (destinationOrder?.id) {
                 this.addPendingOrder([destinationOrder.id]);
             }
-            await this.deleteOrders([order]);
+            if (!isMergedOrder) {
+                order.isTransferedOrder = true;
+                await this.deleteOrders([order]);
+            }
```
[the remark this answers is not in the document]
Comment (caburj) [reply addons/pos_restaurant/static/src/app/services/pos_store.js:410]: alright. I don't like the fact that that method is overloaded. there should be separate method for transfer and for merging.

File: addons/pos_restaurant/static/src/app/services/pos_store.js (hunk 15 above)
Comment (jipr-odoo) [reply addons/pos_restaurant/static/src/app/services/pos_store.js:322]: so should we keep this method as it is for now ?

Comment (caburj) [reply addons/pos_restaurant/static/src/app/services/pos_store.js:322]: yeah, we can keep them.

**Review by caburj — APPROVED** [review]
robodoo r+
